godot_project/editor/controls/dockers/workspace_docker/simulations_docker/simulations_docker_controls/atomic_structure_model_validator.gd

```diff
@@ -126,76 +128,117 @@ func has_overlapping_atoms() -> bool:
-		
-			var original_atoms: PackedInt32Array = nano_structure.get_visible_atoms()
-			var original_positions: PackedVector3Array = PackedVector3Array()
-			for original_atom_id: int in original_atoms:
-				original_positions.push_back(nano_structure.atom_get_position(original_atom_id))
+		var nano_structure: AtomicStructure = overlap.structure_context.nano_structure
+		if not is_instance_valid(nano_structure):
+			continue
 		
-		# Store the atoms positions a small random offset to prevent two atoms having
-		# the exact same coordinates
-		var atoms_positions: Dictionary = {
-			# atom_id<int>: position<Vector3>
-		}
+		# Find the overlap center position.
+		var current_overlap_atoms: Dictionary[int, SpatialHashGrid.Item] = {}
+		var overlap_center: Vector3 = Vector3.ZERO
 		for atom_id in overlap.atoms_id:
 			if not nano_structure.is_atom_valid(atom_id):
-				# atom was deleted after overlaps was validated
+				# Atom was deleted after the validation
 				continue
 			var atom_position: Vector3 = nano_structure.atom_get_position(atom_id)
-			var random_dir: Vector3 = Vector3(randf() - 0.5, randf() - 0.5, randf() - 0.5).normalized()
-			atoms_positions[atom_id] = atom_position + random_dir * MAX_COVALENT_RADIUS * 0.15
+			overlap_center += atom_position
+			static_atom_grid.remove_item_by_position(atom_position)
+			var item: SpatialHashGrid.Item
+			item = overlap_atom_grid.add_item(atom_position, {"id": atom_id, "data": overlap})
+			current_overlap_atoms[atom_id] = item
+
+		if current_overlap_atoms.is_empty():
+			continue
+		overlap_center /= current_overlap_atoms.size()
 		
-		# Relax the atoms positions.
-		# For each atom, find the nearest atom and move away. Repeat up to MAX_SHAKE_ITERATIONS
-		var move_offset: float = MAX_COVALENT_RADIUS * 0.25
-		var max_separation_distance_squared: float = pow(MAX_COVALENT_RADIUS, 2.0)
-		for iteration: int in MAX_SHAKE_ITERATIONS:
-			var atoms_were_moved: bool = false
-			for atom_id: int in atoms_positions:
-				var dir := Vector3.ONE * 99999.0
-				var min_distance := 99999.0
-
-				# Find the closest atom
-				for other_atom_id: int in atoms_positions:
-					if atom_id == other_atom_id:
-						continue
-					var diff: Vector3 = atoms_positions[atom_id] - atoms_positions[other_atom_id]
-					var distance: float = diff.length_squared()
-					if distance < min_distance:
-						dir = diff
-						min_distance = distance
-				
-				if min_distance > max_separation_distance_squared:
-					continue # Atom already far enough
-				
-				atoms_positions[atom_id] += dir.normalized() * move_offset
-				atoms_were_moved = true
+		# Move each overlapping atoms away from the center point, by at least
+		# their atomic radius.
+		for atom_id: int in current_overlap_atoms:
+			var item: SpatialHashGrid.Item = current_overlap_atoms[atom_id]
+			var position: Vector3 = item.position
+			var dir_from_center: Vector3 = overlap_center.direction_to(position)
+			while dir_from_center.is_equal_approx(Vector3.ZERO):
```

Comment:
At first i tought now that you inverted the order of this code `overlap_center.direction_to(position)` would throw an error before this condition is reached, but turns out it doesn't, it just returns `Vector3.ZERO`